### DISCUSSION TITLE
Issue in connecting OPC UA client with Kepware v5 using OPC UA sdk

### DISCUSSION THREAD

**Discussion by prasadmarathe** [eclipse-milo/milo#1123] [General]
Hi Team,
I wanted to use OPCUA SDK to integrate in my java application in order to connect with Kepware server v5. Can you provide me steps to make connection to my kepware server.
I am using below code -
try {
				OpcUaClient client = OpcUaClient.create("opc.tcp://<IP Address>:49320");
				LOGGER.info("Session is created for - " + client.toString());
			} catch (Exception e) {
				LOGGER.log(Level.SEVERE, e.getMessage(), e);
				e.printStackTrace();
			}

**Comment by kevinherron** [comment]
Maybe start here, and make sure your endpoint URL includes the hostname or IP of the KSE instance.

https://github.com/eclipse/milo/wiki/Client#connecting

**Reply by prasadmarathe** [reply]
Hello,
Thank you for providing your help and support.
I am getting below error -
java.lang.AbstractMethodError: io.netty.util.concurrent.MultithreadEventExecutorGroup.newChild(Ljava/util/concurrent/Executor;[Ljava/lang/Object;)Lio/netty/util/concurrent/EventExecutor;

After implementing below code in my existing project.

Implemented code -

try {
	
	OpcUaClient client = OpcUaClient.create("opc.tcp://<IP Address:port>",
			endpoints -> endpoints.stream()
					.filter(e -> e.getSecurityPolicyUri().equals(SecurityPolicy.None.getUri())).findFirst(),
			configBuilder -> configBuilder.build());

	LOGGER.info("Session is created for - " + client.toString());

	client.connect().get();

} catch (Exception e) {
	LOGGER.log(Level.SEVERE, e.getMessage(), e);
	e.printStackTrace();
}

**Please suggest if it is related to milo version. Currently i am using version - 0.6.9**

**Reply by kevinherron** [reply]
Maybe you have an old version of Netty on your class path?

**Reply by prasadmarathe** [reply]
Hello,

Which version of Netty is required for Milo OPC UA client SDK to work.

**Reply by kevinherron** [reply]
4.1.x

**Reply by prasadmarathe** [reply]
Hi, 
I have added dependency in my project as below, but still getting same netty error as mentioned earlier.
<dependency>
    <groupId>org.eclipse.milo</groupId>
    <artifactId>sdk-client</artifactId>
    <version>0.6.9</version>
</dependency>		
<dependency>
    <groupId>io.netty</groupId>
    <artifactId>netty-all</artifactId>
    <version>4.1.77.Final</version>
</dependency>

Is my implementation correct or do i need to add specific artifactId in place of netty-all, for OPC UA JAVA client implementation.

Any sample project of OPC UA Java client which can connect with Kepware OPC server will really help.

**Reply by kevinherron** [reply]
You don't need to add your own Netty dependency, Milo has a transitive dependency.

You need to figure out why your project already has a dependency on an _older_ version of Netty.

**Reply by prasadmarathe** [reply]
Hi Kevin,

After changing netty dependency position and version, my netty issue gets resolved. But now i am getting below error while calling create -
**Code** 
OpcUaClient client = null;
client = OpcUaClient.create("opc.tcp://IP Address:Port number");
				 client.connect().get();
				 System.out.println("Connected to opc.tcp://IP Address:Port number through eclipse Milo ");
				 
**Error**
SEVERE: Error starting wrapper for: OPC(@TST_OPC)
java.lang.NoSuchMethodError: com.google.common.collect.Tables.synchronizedTable(Lcom/google/common/collect/Table;)Lcom/google/common/collect/Table;
	at org.eclipse.milo.opcua.stack.core.types.DefaultDataTypeManager.<init>(DefaultDataTypeManager.java:30)
	at org.eclipse.milo.opcua.stack.core.types.DefaultDataTypeManager.createAndInitialize(DefaultDataTypeManager.java:87)
	at org.eclipse.milo.opcua.stack.client.UaStackClient.<init>(UaStackClient.java:58)
	at org.eclipse.milo.opcua.stack.client.UaStackClient.create(UaStackClient.java:415)
	at org.eclipse.milo.opcua.stack.client.DiscoveryClient.getEndpoints(DiscoveryClient.java:258)
	at org.eclipse.milo.opcua.stack.client.DiscoveryClient.getEndpoints(DiscoveryClient.java:231)
	at org.eclipse.milo.opcua.stack.client.DiscoveryClient.getEndpoints(DiscoveryClient.java:188)
	at org.eclipse.milo.opcua.sdk.client.OpcUaClient.create(OpcUaClient.java:214)


Can you help me identify the root cause here based on above code and error data

**Reply by kevinherron** [reply]
Looks like now your problem is an old version of Guava on the class path.
